src/main/scala/firrtl/passes/InferResets.scala

```diff
@@ -0,0 +1,200 @@
+// See LICENSE for license details.
+
+package firrtl.passes
+
+import firrtl._
+import firrtl.ir._
+import firrtl.Mappers._
+import firrtl.traversals.Foreachers._
+import firrtl.analyses.InstanceGraph
+import firrtl.graph.{DiGraph, MutableDiGraph}
+import firrtl.annotations.{ReferenceTarget, TargetToken}
+import firrtl.Utils.toTarget
+import collection.mutable
+
+import scala.util.{Try, Success, Failure}
+
+/** Infers the concrete type of [[Reset]]s by their connections
+  * This is a global inference because ports can be of type [[Reset]]
+  */
+// TODO should we error if a DefMemory is of type AsyncReset? In CheckTypes?
+object InferResets extends Pass {
+
+  final class DifferingDriverTypesException private (msg: String) extends PassException(msg)
+  object DifferingDriverTypesException {
+    def apply(target: ReferenceTarget, tpes: Seq[(Type, Seq[TypeDriver])]): DifferingDriverTypesException = {
+      val xs = tpes.map { case (t, ds) => s"${ds.map(_.target().serialize).mkString(", ")} of type ${t.serialize}" }
+      val msg = s"${target.serialize} driven with multiple types!" + xs.mkString("\n  ", "\n  ", "")
+      new DifferingDriverTypesException(msg)
+    }
+  }
+
+  private sealed trait ResetDriver
+  private case class TargetDriver(target: ReferenceTarget) extends ResetDriver
+  // We keep the target around (lazily) so that we can report errors
+  private case class TypeDriver(tpe: Type, target: () => ReferenceTarget) extends ResetDriver
+
+  // Collect all drivers for circuit elements of type ResetType
+  private def analyze(c: Circuit): Map[ReferenceTarget, List[ResetDriver]] = {
+    type DriverMap = mutable.HashMap[ReferenceTarget, List[ResetDriver]]
+    def onMod(mod: DefModule): DriverMap = {
+      val instMap = mutable.Map[String, String]()
+      // We need to convert submodule port targets into targets on the Module port itself
+      def makeTarget(expr: Expression): ReferenceTarget = {
+        val target = toTarget(c.main, mod.name)(expr)
+        Utils.kind(expr) match {
+          case InstanceKind =>
+            val mod = instMap(target.ref)
+            val port = target.component.head match {
+              case TargetToken.Field(name) => name
+            }
+            target.copy(module = mod, ref = port, component = target.component.tail)
+          case _ => target
+        }
+      }
+      def onStmt(map: DriverMap)(stmt: Statement): Unit = {
+        stmt match {
+          case Connect(_, lhs, rhs) if lhs.tpe == ResetType =>
+            val target = makeTarget(lhs)
+            val driver = rhs.tpe match {
+              case ResetType => TargetDriver(makeTarget(rhs))
+              case tpe       => TypeDriver(tpe, () => makeTarget(rhs))
+            }
+            map(target) = driver :: Nil
+          case WDefInstance(_, inst, module, _) =>
+            instMap += (inst -> module)
+          case Conditionally(_, _, con, alt) =>
```

Comment:
Does this pass work on High FIRRTL now? If so, you may have SubAccesses in the lhs of the Connect, which I'm not sure Utils.toTarget will work properly.

Reply:
It actually requires `ExpandConnects` so runs during lowering from High to Mid. That being said, I do have it running before `RemoveAccesses` so your point is well taken, I need to test that.

Reply:
This now works on High FIRRTL